agents/src/transcription.ts

```diff
@@ -118,12 +126,72 @@ export class BasicTranscriptionForwarder implements TranscriptionForwarder {
   }
 
   async close(interrupt: boolean): Promise<void> {
-    this.#isRunning = false;
+    this.isRunning = false;
 
     // Publish whatever we had as final
     if (!interrupt) {
-      this.currentCharacterIndex = this.#currentText.length;
+      this.currentCharacterIndex = this.currentText.length;
+    }
+    await this.publishTranscription(true);
+  }
+}
+
+export class ChatAndTranscriptionForwarder extends BasicTranscriptionForwarder {
+  protected type: TranscriptionType;
+  protected originalMessage?: ChatMessage;
+  protected chatMutex: Mutex;
+
+  constructor(
+    room: Room,
+    participantIdentity: string,
+    trackSid: string,
+    messageId: string,
+    transcriptionType: TranscriptionType,
+  ) {
+    super(room, participantIdentity, trackSid, messageId);
+    this.type = transcriptionType;
+    this.chatMutex = new Mutex();
+  }
+
+  override async startPublishingLoop() {
+    this.isRunning = true;
+    let sleepInterval = this.computeSleepInterval();
+    let isComplete = false;
+    while (this.isRunning && !isComplete) {
+      this.currentPlayoutTime += sleepInterval;
+      this.currentCharacterIndex = Math.floor(this.currentPlayoutTime * this.charsPerSecond);
+      isComplete = this.textIsComplete && this.currentCharacterIndex >= this.currentText.length;
+      switch (this.type) {
+        case TranscriptionType.CHAT:
+        case TranscriptionType.CHAT_AND_TRANSCRIPTION:
+          await this.sendChatMessage();
+        case TranscriptionType.TRANSCRIPTION:
+        case TranscriptionType.CHAT_AND_TRANSCRIPTION:
+          await this.publishTranscription(false);
+        default:
+          break;
+      }
+      if (this.isRunning && !isComplete) {
+        sleepInterval = this.computeSleepInterval();
+        await new Promise((resolve) => setTimeout(resolve, sleepInterval * 1000));
+      }
+    }
+
+    if (this.isRunning) {
+      this.close(false);
+    }
+  }
+
+  protected async sendChatMessage() {
```

Comment:
IMO it would be fine to simply truncate at 15KB or up to 63KB and and log a warning, for now.  If we do decide to split into multiple messages things get messy, esp with edits. and tokenization-aware splitting may be tricky too in UTF-8 where characters have variable bytesizes...